src/viam/sdk/common/service_helper.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <type_traits>
+
+#include <viam/sdk/resource/resource_server_base.hpp>
+
+namespace viam {
+namespace sdk {
+
+class ServiceHelperBase {
+   public:
+    ::grpc::Status fail(::grpc::StatusCode code, const char* message) const noexcept;
+
+    ::grpc::Status failNoRequest() const noexcept;
+
+    ::grpc::Status failNoResource(const std::string& name) const noexcept;
+
+    ::grpc::Status failStdException(const std::exception& xcp) const noexcept;
+
+    ::grpc::Status failUnknownException() const noexcept;
+
+   protected:
+    explicit ServiceHelperBase(const char* method) noexcept : method_{method} {}
+
+   private:
+    const char* method_;
+};
+
+template <typename ServiceType, typename RequestType>
+class ServiceHelper : public ServiceHelperBase {
+   public:
+    ServiceHelper(const char* method, ResourceServer* rs, RequestType* request) noexcept
+        : ServiceHelperBase(method), rs_{rs}, request_{request} {};
+
+    template <typename Callable>
+    ::grpc::Status operator()(Callable&& callable) const noexcept try {
+        if (!request_) {
+            return failNoRequest();
+        }
+        const auto resource = rs_->resource_manager()->resource<ServiceType>(request_->name());
+        if (!resource) {
+            return failNoResource(request_->name());
+        }
+        return invoke_(std::forward<Callable>(callable), std::move(resource));
+    } catch (const std::exception& xcp) {
+        return failStdException(xcp);
+    } catch (...) {
+        return failUnknownException();
+    }
+
+    auto getExtra() const {
+        return request_->has_extra() ? struct_to_map(request_->extra()) : AttributeMap{};
+    }
+
+   private:
+    template <typename Callable, typename... Args>
+    using is_void_result = std::is_void<std::result_of_t<Callable(Args...)>>;
+
+    // Implementation of `invoke_` for a Callable returning non-void,
+    // presumably an error return, which we return as a
+    // ::grpc::Status.
+    template <typename Callable,
+              typename ResourcePtrType,
+              std::enable_if_t<!is_void_result<Callable, ServiceHelper&, ResourcePtrType&&>::value,
```

Comment:
This is so cool 🧑‍🔧 .

Reply:
It is much easier with Concepts but this is as good as it gets in C++14. But yeah, SFINAE is sort of magical.

Reply:
Yeah semantic categories are hugely useful. I was initially surprised when `template` didn't really (seem to) offer that.